### PR TITLE
feat(backend): function to add hidden dapps

### PR DESCRIPTION
**Pull request by AntonioVentilii**
# Motivation

We want to add an endpoint to add dApps that the user wants hidden in the dApps Carousel.

# Changes

- Add function `add_hidden_dapp_id` to `user_profile` implementation.
- Create error types and returns types
- Propagate the function up to open endpoint in `lib`

# Tests

Included tests for new endpoint.

### CODE REVIEW

File: src/shared/src/impls.rs [hunk 1]
```diff
@@ -164,6 +164,35 @@ impl StoredUserProfile {
         new_profile.updated_timestamp = now;
         Ok(new_profile)
     }
+
+    /// # Errors
+    ///
+    /// Will return Err if there is a version mismatch or if the dApp ID is already hidden.
```
Comment (bitdivine) [inline src/shared/src/impls.rs:170]: Personally I wouldn't return an error if the app is already hidden.  It is often helpful if functions are idenpontent, so saying "hide x!  hide x!  hide x!" should get the answer "ok, ok, ok, calm down!".  But it's a matter of taste.  Up to you.

File: src/backend/src/lib.rs [hunk 2]
```diff
@@ -518,6 +520,29 @@ pub fn add_user_credential(
...
 }
 
+/// Adds a dApp ID to the user's list of hidden dApp IDs.
+///
+/// # Errors
+/// - Returns `Err` if the ID is already in the `hidden_dapp_ids`, or the user profile is not found, or the user profile version is not up-to-date.
+#[update(guard = "may_write_user_data")]
+pub fn add_user_hidden_dapp_id(
+    request: AddHiddenDappIdRequest,
+) -> Result<(), AddDappSettingsError> {
```
Comment (bitdivine) [inline src/backend/src/lib.rs:530]: Can I provide a 1Mb long app ID?  Sounds as if I might be able to do a DOS fairly cheaply.  See the comment at the end of the PR for a suggested sanity check to call.

**Review by bitdivine — APPROVED** [review]

File: src/shared/src/impls.rs (hunk 1 above)
Comment (AntonioVentilii) [reply src/shared/src/impls.rs:170]: eheh agreed! changed

File: src/backend/src/user_profile.rs [hunk 3]
```diff
@@ -51,3 +52,17 @@ pub fn add_credential(
         Err(AddUserCredentialError::UserNotFound)
     }
 }
+
```
[the remark this answers is not in the document]
Comment (AntonioVentilii) [reply src/backend/src/user_profile.rs:55]: right! included

File: src/backend/src/lib.rs (hunk 2 above)
Comment (AntonioVentilii) [reply src/backend/src/lib.rs:536]: added a test for the max length of the ID

shall I add something for the max number of IDs that can be saved?

Comment (bitdivine) [reply src/backend/src/lib.rs:536]: Sounds like a good idea. :-) But doesn't have to be in this PR.

File: src/backend/src/user_profile.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bitdivine) [reply src/backend/src/user_profile.rs:55]: Thanks!

File: src/shared/src/impls.rs (hunk 1 above)
Comment (bitdivine) [reply src/shared/src/impls.rs:170]: Thanks!

**Review by bitdivine — APPROVED** [review]
LGTM, thanks!

File: src/backend/src/lib.rs (hunk 2 above)
Comment (AntonioVentilii) [reply src/backend/src/lib.rs:536]: Done with:

- https://github.com/dfinity/oisy-wallet/pull/5248

**Pull request merged**